File: sort/bucket_sort/Java/BucketSort.java
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.IntStream;

public class BucketSort {
    public static void main(String[] args) {
        int[] elements = init();
        sort(elements);
        System.out.println(Arrays.toString(elements));
    }

    public static void sort(int[] elements) {
        int[] code = hash(elements);
        List<Integer>[] buckets = new List[code[1]];

        IntStream.range(0, buckets.length).forEach(i -> buckets[i] = new ArrayList());

        Arrays.stream(elements).forEach(i -> buckets[hash(i, code)].add(i));
        Arrays.stream(buckets).forEach(bucket -> Collections.sort(bucket));

        final AtomicInteger idx = new AtomicInteger(0);
        Arrays.stream(buckets).forEach(bucket -> {
            for(int el : bucket) {
                elements[idx.getAndIncrement()] = el;
            }
        });
    }

    private static int[] hash(int[] elements) {
        int max = Arrays.stream(elements).reduce(Integer.MIN_VALUE, (result, current) -> Math.max(result, current));
        return new int[] { max, (int) Math.sqrt(elements.length) };
    }

    private static int hash(int i, int[] code) {
        return (int)((double) i / code[0] * (code[1] - 1));
    }

    private static int[] init() {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the number of elements: ");
        int n = sc.nextInt();
        int[] elements = new int[n];
        System.out.println("Enter elements: ");
        for(int i = 0; i < n; i++)
            elements[i] = sc.nextInt();
        return elements;
    }
}

File: sort/bucket_sort/C#/BucketSort.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    class BucketSort
    {
        static int[] Sort(int[] numbers)
        {
            int minValue = numbers[0];
            int maxValue = numbers[0];

            for (int i = 1; i < numbers.Length; i++)            
            {
                if (numbers[i] > maxValue) maxValue = numbers[i];
                if (numbers[i] < minValue) minValue = numbers[i];
            }

            List<int>[] buckets = new List<int>[maxValue - minValue + 1];

            for (int i = 0; i < buckets.Length; i++)
            {
                buckets[i] = new List<int>();
            }

            for (int i = 0; i < numbers.Length; i++)
            {
                buckets[numbers[i] - minValue].Add(numbers[i]);
            }

            int k = 0;

            for (int i = 0; i < buckets.Length; i++)
            {
                if (buckets[i].Count > 0)
                {
                    for (int j = 0; j < buckets[i].Count; j++)
                    {
                        numbers[k] = buckets[i][j];
                        k++;
                    }
                }
            }

            return numbers;
        }

        static void Main(string[] args)
        {
            int[] numbers = { 1, 9, -4, 15, 79, 2, 4, -10 };
            numbers = Sort(numbers);

            foreach (var number in numbers)
            {
                Console.WriteLine(number);
            }
        }
    }
}
